Fix and test for include_total not working

davidread opened the pull request:
include_total wasn't working because of the default() validator was broken, so I fixed that.

Fixes #4446

### Proposed fixes:



### Features:

- [x] includes tests covering changes
- [ ] includes updated documentation
- [ ] includes user-visible changes
- [ ] includes API changes
- [ ] includes bugfix for possible backport

Please [X] all the boxes above that apply

wardi commented on ckanext/datastore/tests/test_search.py:
```diff
@@ -634,6 +634,18 @@ def test_search_is_unsuccessful_when_called_with_invalid_fields(self):
+        assert res_dict['success'] is True
+        result = res_dict['result']
+        assert 'total' not in result
+
```
Would you add this to the class at the top instead of with the legacy tests and like those tests use `helpers.call_action` to skip the request processing bits.

wardi commented on ckanext/datastore/controller.py:
```diff
@@ -143,7 +143,7 @@ def result_page(offs, lim):
             'sort': '_id',
             'records_format': records_format,
-            'include_total': 'false',  # XXX: default() is broken
+            'include_total': 'false',
```
we can change this to `False` now :raised_hands:

davidread commented:
@wardi thanks for the review. I've made those two improvements now :)

davidread commented on ckan/lib/navl/validators.py:
```diff
@@ -78,7 +78,7 @@ def default(default_value):
     def callable(key, data, errors, context):
```
Good idea. Added!

davidread commented:
I've added these things to the changelog.

I've also removed an unnecessary total/`SELECT COUNT(*)` call by `resource_view_get_fields()`. I think it hangs well with this same PR since it is the same topic and is safe one-liner.

This PR now gets rid of 2 of the 4 `SELECT COUNT(*)` calls done when the user views the resource page. Which should make a difference to server load for large resources (e.g. 1 million rows / 100MB +) - see https://github.com/ckan/ckanext-xloader/issues/41

wardi commented on CHANGELOG.rst:
```diff
@@ -10,6 +10,16 @@ Changelog
+Minor changes:
+
+ * For navl schemas, the 'default' validator no longer applies the default when
+   the value is False, (), [] or {} (#4448)
```
please remove `()` (shouldn't appear as part of a json payload) and add `0` to this list.

wardi commented on ckan/lib/navl/validators.py:
```diff
@@ -74,6 +74,8 @@ def ignore(key, data, errors, context):
 
 def default(default_value):
+    '''When key is missing or value is 'empty', set it with a default value.
+    See the code for how 'empty' is defined.'''
```
I don't think it hurts to repeat ourselves here.

`When key is missing or value is an empty string or None, replace it with a default value`

One day we could include the validator docstrings in the proper docs.

wardi commented on ckan/lib/helpers.py:
```diff
@@ -2280,7 +2280,8 @@ def resource_view_get_fields(resource):
         'resource_id': resource['id'],
-        'limit': 0
+        'limit': 0,
+        'include_total': False,
```
looks good!

davidread commented:
Comments addressed!
